server/src/main/java/com/paypal/selion/grid/AbstractBaseLauncher.java

```diff
@@ -0,0 +1,408 @@
+
+    /*
+     * Whether this launcher has been initialized.
+     */
+    private boolean initialized = false;
+
+    /*
+     * the current instance type
+     */
+    private InstanceType type;
+
+    /*
+     * The received launcher commands
+     */
+    private List<String> commands;
+
+    public final boolean isInitialized() {
+        return initialized;
+    }
+
+    /**
+     * Set this launchers to initialized
+     * 
+     * @param state
+     *            the new state
+     */
+    final void setInitialized(boolean state) {
+        this.initialized = state;
+    }
+
+    /**
+     * Set this launchers commands
+     * 
+     * @param commands
+     *            the launcher commands
+     */
+    final void setCommands(List<String> commands) {
+        this.commands = commands;
+    }
+
+    /**
+     * @return the launcher commands
+     */
+    final List<String> getCommands() {
+        return commands;
+    }
+
+    /**
+     * Get the {@link InstanceType} this launcher represents. If type has not been established by a previous call to
+     * {@link #setType(InstanceType)} this method attempt to determine the type by inspecting the {@link #commands}.
+     * Inspection does not work for mobile instance types. Therefore, it is best to always call
+     * {@link #setType(InstanceType)} beforehand.
+     * 
+     * @return an {@link InstanceType}
+     */
+    final InstanceType getType() {
+        if (type == null) {
+            return determineType(commands);
+        }
+        return type;
+    }
+
+    /**
+     * Set the {@link InstanceType} for this launcher.
+     * 
+     * @param type
+     *            the {@link InstanceType}
+     */
+    final void setType(InstanceType type) {
+        this.type = type;
+    }
+
+    /**
+     * Return the {@link InstanceType}. Defaults to {@link InstanceType#SELENIUM_STANDALONE}. Does not work or mobile
+     * instance types.
+     * 
+     * @param commands
+     *            command line arguments passed from User
+     * @return an {@link InstanceType}
+     */
+    private InstanceType determineType(List<String> commands) {
+        LOGGER.entering(commands.toString());
+        type = InstanceType.SELENIUM_STANDALONE;
+        if (commands.contains(ROLE_ARG) && commands.contains(InstanceType.SELENIUM_NODE.getFriendlyType())) {
+            type = InstanceType.SELENIUM_NODE;
+        }
+        if (commands.contains(ROLE_ARG) && commands.contains(InstanceType.SELENIUM_HUB.getFriendlyType())) {
+            type = InstanceType.SELENIUM_HUB;
+        }
+        LOGGER.exiting(type);
+        return type;
+    }
+
+    /**
+     * Get program arguments to pass
+     *
+     * @return the program arguments to pass represented as an array of {@link String}
+     * @throws IOException
+     */
+    String[] getProgramArguments() throws IOException {
+        LOGGER.entering();
+        List<String> args = new LinkedList<String>();
```

Comment:
Do we need to iterate through every element for this initialization ?

Can we do something like below ? 

`List<String> args = new LinkedList<String>(commands);`